Update the CGNS reader to work with multiblock visit files.

brugger1 opened the pull request:
### Description

I enhanced the CGNS reader so that it works with multiblock visit files. The reader was strictly MTMD and I modified it to work either as MTSD or MTMD. In normal usage it returns an MTMD database and in the case of a visit file with an NBLOCKS not equal to 1 it returns an MTSD database. It only handles the case where each CGNS file has a single block and the number of variables and number of time states is the same for all of them. This is not checked, since that would require a large overhead on open.

This change is mostly ready to go, but I wanted to get input on the changes at this point. I still need to update the release notes and I need to chase down a core dump at exit.

### Type of change

- [X] New feature

### How Has This Been Tested?

I got a collection of CGNS files where each CGNS file had a single block. I was able to open and plot the mesh and variables.

### Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] I have updated the release notes
- [ ] I have made corresponding changes to the documentation
- [ ] I have added debugging support to my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added any new baselines to the repo
- [X] I have assigned reviewers

brugger1 committed:
Update the CGNS reader to work with multiblock visit files.

markcmiller86 commented on src/databases/CGNS/CGNSCommonPluginInfo.C:
```diff
@@ -20,7 +38,7 @@
 CGNSCommonPluginInfo::GetDatabaseType()
 {
-    return DB_TYPE_MTMD;
+    return dbType;
```
Can `GetDatabaseType()` be called before `SetupDatabase()`. Maybe it *can* be but VisIt doesn't wind up operating that way? Reason I ask is that we don't actually know database type until after `SetupDatabase()` so returning a known type (the default) instead of something that indicates that the type is unknown might be a little fib.

markcmiller86 replied:
Actually, I think this is a non-issue. I went looking around for where `GetDatabaseType()` is actually used and I didn't find a single case where that method is called...

```
[scratlantis:~/visit/visit/src] miller86% find . -name databases -prune -false -o \
-name '*.[Ch]' -exec grep 'GetDatabaseType()' {} \; -print
            h << "    virtual DatabaseType              GetDatabaseType();" << endl;
./tools/dev/xml/GenerateInfo.h
    virtual DatabaseType              GetDatabaseType() = 0;
./common/plugin/DatabasePluginInfo.h
`

markcmiller86 commented on src/databases/CGNS/CGNSCommonPluginInfo.C:
```diff
@@ -44,14 +62,42 @@ avtDatabase *
 CGNSCommonPluginInfo::SetupDatabase(const char *const *list,
                                    int nList, int nBlock)
 {
+    avtDatabase *db = NULL;
```
I can't recall if this is ever an issue but if xml2info is used to regen, will this custom code survive? I also can't recall if we added ability to indicate the existence of *custom* code in generated files in the `.xml` file to warn about this. If so, I didn't see that flag in the `.xml` file.

The above may all be me blowing smoke and confusing regular *attribute* objects (which are often re-gen'd) with database plugins which maybe are rarely if at all.

brugger1 replied:
I believe this falls under the use case of generating it once and not modifying it again. This is done in other readers, so I will go with precedence on this.

markcmiller86 reviewed (changes requested):
So in summary, the bulk of the original plugin goes to `avtCGNSFileReader` leaving the original plugin to be a pair of small wrappers around `avtCGNSFileReader`; one for the MTMD case and one for the MTSD case. In the SD case, the domain number passed to the reader is always zero. I like it :heart:. Seems like a decent/clean solution here. Did you have any thoughts for a possible future generalization above *all* plugins so that this doesn't wind up having to get duplicated in other cases it might be needed (we appear to have 32 MTMD plugins).

I know it doesn't make sense to confirm all files have single mesh and same number of timesteps during open/popmd, but what about issuing error from engine if that condition is later found to be false. I dunno if that *requires* an additional all-reduce there and if so, forget about it. I am just thinking that at some point, VisIt would know that these original assumptions have been violated and it would be nice to warn the unsuspecting user. That icing though.

brugger1 committed:
Add a check for a single mesh in the MTSD case.

brugger1 committed:
Update the release notes.

brugger1 commented:
I agree, doing this so that it can apply to all MTMD databases would be great. It seems like that should be possible, since doing for each one individually on an as needed basis is ridiculous.

I added a check in PopulateDatabaseMetaData that there is a single block and throw an invalid file exception if there is more than one mesh. I also updated the release notes.

brugger1 committed:
Merge branch '3.1RC' into task/brugger1/2020_02_28_cgns_reader

markcmiller86 commented on src/resources/help/en_US/relnotes3.1.2.html:
```diff
@@ -33,6 +33,7 @@
 <ul>
   <li>Removed the pbatch launch profile from the Lawrence Livermore National Laboratory's Max host profile.</li>
   <li>Enhanced the Pick Through Time to handle vectors, tensors, and arrays when picking original data. For vectors, the magnitude is used. For tensors and arrays, the major eigenvalue is used.</li>
+  <li>Enhanced the ".visit" file reader to support grouping multiple single-block CGNS files into a multi-block file. The reader requires that each CGNS file contain a single domain. The ".visit" file will need a "!NBLOCKS ###" directive as the first line in the file, followed by all the CGNS filenames, one filename per line. You may also need to specify the file type as CGNS when opening the fil ...
```
There isn't really a `.visit` file reader. The `.visit` file concept lives *above* any specific SD plugin. Also,  CGNS community uses *block* for *domain*. So, I would word this a little differently.

> Enhanced the CGNS plugin to support a mesh decomposed across multiple CGNS files, one *block* per file as might be typical of file-per-processor parallel I/O. This requires that each file contain just a *single* CGNS mesh block. In addition, if all files do not also contain the same number of time steps, results will be indeterminate. Finally, the ensemble of CGNS files must be grouped together into a [.visit](https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/gui_manual/WorkingWithFiles/Supported_File_Types.html?highlight=nblock#creating-visit-files) file.

brugger1 replied:
I will change the release notes on another commit. I tried to do it on the existing branch and got an error.

markcmiller86 reviewed (approved):
Looks good. Good idea to add a check to popmd.

I suggested a change to the release notes.
